### ISSUE TITLE
questions about the model and the source code

### ISSUE THREAD

**Issue by xvii129** [xupei0610/SocialVAE#2]
1. 双向RNN考虑了整条轨迹，反向RNN对未来轨迹点的使用是否有数据泄露？模型推理时是否依赖完整轨迹（观察轨迹+预测轨迹）
2. 源码中并不支持只输入obs_horizon长度数据进行预测（obs_horizon+pred_horizon长度的数据可处理），模型载入使用时有什么要注意的吗？怎么理解源码中prediction = prediction + input[-1, ..., :2]在输入数据上直接+偏移得到预测数据的操作

i.e.
1. bidirectional RNN structure takes the whole ground-truth trajectory into account, whether there is data leakage in the use of future trajectory points by the backward RNN, during inference, whether a complete trajectory(observation points and prediction points) is required?
2. in the source code, prediction of input x with only obs_horizon data points seems not supported, but input with obs_horizon and prediction horizon points can be processed normally, are there any requirements for model inference? what's more, prediction trajectory points are obtained from this code "prediction = prediction + input[-1, ..., :2]", how to understand the operation of directly adding displacement offset to the whole input data

**Comment by xupei0610** [comment]
1. 不是。反向RNN只用于后验估计。在推理时，是不需要未来轨迹的数据做反向RNN的。可以参见https://github.com/xupei0610/SocialVAE/blob/de4d54f2ec8c818a6e5cb016fdb0b8eb6cdb3264/main.py#L172-L187
在推理时，ground-truth值y并没有被用于推理，只被用于推理之后计算ADE/FDE。

No. The backward RNN is just used for posterior approximation. During inference, there is no need of any future trajectory data. The backward RNN will not run during inference actually. The given code can be used as a reference, where we can see that the ground-truth value of the future trajectory data (`y`) is only used for ADE/FDE computation after model inference.


2.1. 预测模型的ob_horizon和pred_horizon是在训练时确定的。参见位于config文件夹中的配置文件。也可以尝试在模型训练后的测试阶段使用不同的ob_horizon和pred_horizon，但是性能可能会有所下降。

The hyperparameters of `ob_horizon` and `pred_horizon` are decided right before training. Their definitions are given by the configuration files in the folder of `config`.  It is also possible to try different configurations during the training and evaluation phases. But there may be a performance decrease.

2.2 模型本身输出的是被预测对象从一帧到下一帧的位移。因此，在获得最终的,全局坐标的运动轨迹时，需要加上观察到的最后一帧的具体位置。

The direct output of our model is the position displacement between two frames, instead of the actual global coordinates. Therefore, to get a whole trajectory in the coordinate of the global, world space, the final output should be added with the agent positions at the last frame of observation.

**Comment by xvii129** [comment]
I appreciate your reply and the article inspired me a lot~ but some questions remain unclear for me: 
1. The backward RNN is just used for posterior approximation, but the latent variable is drawn implicitly by both forward and backward state during training, then whether there is data leakage when using the whole trajectory as training features instead of results for comparison? 
2. obs_horizon and pred_horizon values are given by the configuration files in the folder of config, suppose I set obs_horizon = 8 and pred_horizon =12, the source code seems not support prediction of input with only obs_horizon = 8 points, even though input with 20 points(the whole trajectory), here is another problem: prediction results are obtained by these two lines "pred = torch.cumsum(d, 0); pred = pred + x[, ..., :]", note that x[, ..., :] contains not only observation points but also prediction points, in that case output is obtained by adding displacement offset to x directly, I'm confused about this.

The above are questions I have when trying to test the model, maybe I didn't understand the article correctly, I would appreciate it if you could give more guidance, thanks a lot

**Comment by xupei0610** [comment]
There is absolutely **NO** data leakage. The groud-truth future trajectories do not participate in the prediction at all.

According to your description, I do not get your understanding of 'data leakage'. What do you mean by "there is data leakage when using the whole trajectory as traing features"? During training, all data are employable in any way you like. There is no data leakage, as long as no ground-truth target data are used for inference during evaluation/testing. That is how training is done in a supervised way.

Generally, during training, VAE draws latent samples from the posterior (approximated based on given ground-truth values), and, is optimized to minimize the KL divergence between the posterior and prior (which is often assumed as normal distributions). During prediction, there is no posterior at all, because no groud-truth data will be given, and latents are drawn from the prior.

Referring to Section 3.A. in our paper, SocialVAE has two sub-models called `Generative Model` and `Inference Model`, where `Generative Model` is corresponding to the prediction phrase and `Inference Model` is corresponding to the learning/training phase. In `Inference Model` (learning/training phrase), latents (`z`) are drawn from `q_\xi` where the backward RNN involves the future, ground-truth trajectories (`b`).  But, in `Generative Model` (prediction phrase), latents (`z`) are drawn from `p_\phi`, where only a forward RNN participates. I would also refer to Fig. 1. in our paper, features from future trajectories only participate in the training phase (red parts). There are no features from future trajectories when doing predictions during evaluation.

From the code level, in the code I shared, you can see that when doing prediction, `y` (the ground-truth future trajectories/the prediction target) is even not given to the model at all. How will the data leak? You can also double-check the `forward` function in `social_vae.py`. No ground-truth future trajectories will be used during prediction.


As for your 2nd question, quick answer: during the learning/training phase, you can give input (observation and target future trajectories) with any length, not even the same with your configuration; during prediction, you can give observation with any length, while the model will return predictions with 12 waypoints (assuming you set PRED_HORIZON=12 in your configuration file).

I would suggest you try `print(x.shape)` in the `forward` function of `social_vae.py` to first verify your guess that the input must have 20 waypoints. It should output `8x128x6` as a tensor with OB_HORIZON=8 and BATCH_SIZE=128. The last dimension `6` is corresponding to global positions (2), velocity (2), and acceleration (2). Further, try to change the configuration file and see if the shape of `x` changes correctly.
You can also set `x=x[0:1]`, as an example, to see if the model works, regardless of the performance, when the given input just has 1 waypoint.
I think you may be unclear what is the output of data loader. It outputs `x` (observation in the shape of OB_HORIZON x BATCH_SIZE x 6), `y` (prediction target, or say ground-truth future trajectories in the shape of PRED_HORIZON x BATCH_SIZE x 2) and `neighbor` (neighbor agents' and velocity and acceleration in the shape of OB_HORIZON+PRED_HORIZON x BATCH_SIZE x 6). Please note that though `neighbor` will have data from both OB_HORIZON and PRED_HORIZON, the PRED_HORIZON part will be ignored if `y` is not provided (see `enc` function in `social_vae.py`) and thus it is safe to provide `neighbor` as OB_HORIZON x BATCH_SIZE x 6 for prediction. So, there is not any data leakage during the prediction phase (calling of the `forward` function in `social_vae.py`), since only `x` and `neighbor` are given to the function. 

For the 3rd question, `x` is the observation (global coordinates), and `pred` is the output (displacement) from the model. So, to get the global coordinates of the prediction, it needs to add `pred` to the global coordinates at the last frame of observation.

**Comment by xvii129** [comment]
okay, thank you for your detailed explanation~

I did evaluate the pretrained model and also fully trained a model according to the tutorial you provided, the 2nd and 3rd questions are encountered during model inference after loading a trained model. 

to put it another way, let's focus on the forward function in the social_vae.py, assuming that input x.shape = 1 x 20 x 6 (20 frames for the whole trajectory and no frameskip, 6 is corresponding to positions/velocity/acceleration) and without neighbors, then we got displacement d.shape = 12 x 20 x 2, 12 for pred_horizon = 12 future frames, 2 for x/y, how to understand the dimension 20 (=8+12, and it will change when setting different obs_horizon and pred_horizon), that's **question 1**. 

based on the above results, via pred = torch.cumsum(d, 0) we got pred.shape = 12 x 20 x 2 as well, to simplify the description, we assume n_predictions = 0 then pred = pred + x[-1, ..., :2], x[, ..., :2].shape = 1 x 20 x 2 including the whole trajectory with observation points and prediction points, so here comes **question 2**: by adding displacement to all trajectory points, for example, there is a trajectory with 8 obs points and 12 pred points, by adding displacement to each frame point coordinate. what's the purpose of this operation, I don't quite understand the practical physical meaning or how can I get prediction trajectory by pred results?  

sorry for the delay of your time, maybe it's because my previous description is not clear enough, please reply when you are available, thanks

**Comment by xupei0610** [comment]
The input data are `batch_first=False`, i.e. the first dimension are the time dimension. The 2nd dimension indicates the number of instances that is going to do prediction, i.e. batch size. 
So, it is `20 x 1 x 6` if there is only one trajectory and the trajectory is 20-frame long. 
If `x` is in the shape of `1 x 20 x 6`, it means a batch of 20 trajectories, each of which only has 1 frame of waypoint. And, finally, you will get a prediction tensor in the shape of `12 x 20 x 2`. It means the predictions for the given 20 trajectories, each of which is 12-frame long.
I think that is where your confusion is from.

And, `x` is not the whole trajectory. It is only the observation. Ground-truth future trajectories are always denoted by `y` in the code, which is not given to the `forward` function. (It is given to the `loss` function for training purposes.)

For clarification, `displacement` means the position difference of an agent between two frames, or in Chinese 被预测对象在两帧间的位移（坐标差）. 
`pred = torch.cumsum(d, 0)` means that we cumulatively add the displacement along the time dimension (0 indicates the 1st dimension) for each trajectory. 
So, finally, we need to add `pred` to `x[-1,...,:2]` (the position of N instances in the last (-1) frame of observation) to get the global coordinates of the predicted trajectories.
Or, in Chinese, 需要再加上初始位置（即，被预测对象在被观察到的最后一帧的位置）以获得完整的预测轨迹坐标.

**Comment by xvii129** [comment]
the explanation about the dimensions is quite helpful, the meanings of displacement and torch.cumsum(d, 0) operation are already clear for me actually, haha~ now I have a better understanding of the article and I would like to change the data for further test as you described, maybe feedback to you in the future. thanks again, best regards.
